### DISCUSSION TITLE
Start app if it doesn't have a window or activate window if it exists

### DISCUSSION THREAD

**Discussion by sommerper** [YaLTeR/niri#267] [Q&A] [answered]
Hi!

On my Pop_os installation I have a script that I bind to shortcuts for manipulating my most used apps like browser, terminal, and code editor.

For the terminal it would look like this:
```
mod+ctrl+shift+n      open or go to terminal
mod+ctrl+shift+alt+n  open or move terminal to current workspace
```

Let's say I have no apps open:

- If I press `mod+ctrl+shift+n` the terminal will be loaded and activated on the current workspace.
- If I'm using the browser on another workspace and press `mod+ctrl+shift+n` I will be taken to the workspace of the terminal and the terminal window will be activated.
- If I'm on another workspace than the terminal and press `mod+ctrl+shift+alt+n` the terminal window will be brought to the current workspace and activated. If it does not exist it will be loaded instead.

Is there a way to check for open apps and their windows and is it possible to manipulate the windows position and workspace location by script?

I hope I made myself clear! It's not that easy to explain.

Thanks!

**Comment by YaLTeR** [comment, marked as answer]
Unfortunately, it's not currently possible. The IPC system needs to be extended with workspace and window information, and some way to manipulate windows different from the currently focused one.

**Reply by sommerper** [reply]
Ok, I understand. I wish I could help but I hardly know any Rust.
Thanks for a quick reply!

**Reply by sommerper** [reply]
I'll see if I can live without it - the workflow is so different I might not even need it anymore. :-)

**Comment by Brisingr05** [comment]
There's a tool called [wlrctl](https://git.sr.ht/~brocellous/wlrctl) that you could check out. I have a keybinding that a runs a script with `wlrctl window focus firefox || firefox` in it. This will go to Firefox if it already exists, otherwise it'll open it.

**Reply by YaLTeR** [reply]
Indeed, that could work.

**Reply by sommerper** [reply]
Thanks for the tip, this does indeed work!

I think with a little bit of scripting I might also be able to pull a window to the current workspace. Then I could switch to Niri fulltime - which is awesome.

**Reply by Schweber** [reply]
@sommerper If you manage to do that, please let me know. I'm looking to do [something similar](https://github.com/YaLTeR/niri/discussions/46).

**Reply by sommerper** [reply]
@Schweber : What @Brisingr05 mentioned works pretty great to be honest.

`wlrctl window focus firefox` brings me to Firefox even on another workspace - although without animation. 

I thought I was able to figure something out so I could bring an app to my current workspace by activating it and then doing something like `move-column-to-workspace` - but currently the IPC doesn't have a way to get the workspace number. It's not something I use that often so I can live with that for now and then, hopefully, one day, the IPC will support it.

Right now my script is pretty simple:

```
#/usr/bin/bash

# TODO: fetch app
if [ "$1" = "fetch" ]; then
    # exit
    exit
fi

# goto app
if [ "$1" = "goto" ]; then
    wlrctl window focus $2 || niri msg action spawn -- $3
    exit
fi
```

And then I can call it with:
```
Mod+Ctrl+Shift+N { spawn "/home/myname/scripts/niri_win.sh" "goto" "org.wezfurlong.wezterm" "/home/myname/bin/wezterm --config enable_wayland=true"; }
Mod+Ctrl+Shift+E { spawn "/home/myname/scripts/niri_win.sh" "goto" "code-url-handler" "/usr/share/code/code --enable-features=UseOzonePlatform,WaylandWindowDecorations --ozone-platform-hint=auto --unity-launch"; }
Mod+Ctrl+Shift+I { spawn "/home/myname/scripts/niri_win.sh" "goto" "vivaldi-stable" "/usr/bin/vivaldi-stable --enable-features=UseOzonePlatform --ozone-platform=wayland"; }
```

Compare that to my hyprland version:

```
#!/usr/bin/bash

# get the active window with hyprctl
active_window_json=$(hyprctl activewindow -j)

# get the window address
window_address=$(echo $active_window_json | jq -r '.address')

# get the window workspace
window_workspace=$(echo $active_window_json | jq -r '.workspace')

# get the workspace id
workspace_id=$(echo $window_workspace | jq -r '.id')

# next workspace id is the current workspace id + 1
next_workspace_id=$(($workspace_id + 1))
prev_workspace_id=$(($workspace_id - 1))

# echo $1
if [ $1 = "movewintonextworkspace" ]; then
    # echo "move to next"
    hyprctl dispatch movetoworkspace "$next_workspace_id,address:$window_address"
    # echo "move to next" >> hypr.log
fi

if [ $1 = "movewintoprevworkspace" ]; then
    echo "move to prev"
    # if prev_workspace_id is 0 do nothing
    if [ $prev_workspace_id -eq 0 ]; then
        exit 0
    fi
    hyprctl dispatch movetoworkspace "$prev_workspace_id,address:$window_address"
fi

if [ $1 = "goto" ]; then
    echo "going to"
    # get all clients
    all_clients=$(hyprctl clients -j)
    wanted_title=$2
    echo "wanted title: $wanted_title"
    # get the client with the class $wanted_class
    
    wanted_client=$(echo $all_clients | jq -r ".[] | select(.initialTitle == \"$wanted_title\")")

    if [ -z "$wanted_client" ]; then
        echo "no client found with title $wanted_title"
        wanted_class=$2
        wanted_client=$(echo $all_clients | jq -r ".[] | select(.initialClass == \"$wanted_class\")")
        
    fi

    if [ -z "$wanted_client" ]; then
        echo "no client found with class $wanted_class"
        echo "opening $3"
        hyprctl dispatch exec $3
        exit 0
    fi

    echo $wanted_client
    client_window_address=$(echo $wanted_client | jq -r '.address')
    # get the window workspace
    client_window_workspace=$(echo $wanted_client | jq -r '.workspace')
    # get the workspace id
    client_workspace_id=$(echo $client_window_workspace | jq -r '.id')
    
    hyprctl dispatch workspace $client_workspace_id
    hyprctl dispatch focuswindow address:$client_window_address
fi

if [ $1 = "fetch" ]; then
    echo "fetching"
    # get all clients
    all_clients=$(hyprctl clients -j)
    wanted_title=$2
    echo "wanted title: $wanted_title"
    # get the client with the class $wanted_class
    
    wanted_client=$(echo $all_clients | jq -r ".[] | select(.initialTitle == \"$wanted_title\")")

    if [ -z "$wanted_client" ]; then
        echo "no client found with title $wanted_title"
        wanted_class=$2
        wanted_client=$(echo $all_clients | jq -r ".[] | select(.initialClass == \"$wanted_class\")")
        
    fi

    if [ -z "$wanted_client" ]; then
        echo "no client found with class $wanted_class"
        echo "opening $3"
        hyprctl dispatch exec $3
        exit 0
    fi

    client_window_address=$(echo $wanted_client | jq -r '.address')
    # get the window workspace
    client_window_workspace=$(echo $wanted_client | jq -r '.workspace')
    # get the workspace id
    client_workspace_id=$(echo $client_window_workspace | jq -r '.id')
    
    hyprctl dispatch movetoworkspacesilent "$workspace_id,address:$client_window_address"
    hyprctl dispatch focuswindow address:$client_window_address
    
fi
```

**Reply by Schweber** [reply]
@sommerper Thanks! 

As for `Hyprland`, i'll do a shameless plug and refer you to [hdrop](https://github.com/Schweber/hdrop). I'm hoping to implement the same for `niri` eventually and use it as my daily driver.

**Reply by sommerper** [reply]
@Schweber Thanks for the tip! That could have saved me the work of doing my own script and yours looks better.
